Implement 'findPattern'

TheGreatMcPain opened the pull request:
CrewLink 2.0.0 now requires 'findPattern' in order to function.

Currently I've just added the pattern.cc and pattern.h files, and modified
them a bit.

TheGreatMcPain committed:
Add WIP pattern.{cc,h}

TheGreatMcPain committed:
Copy and use memory::readMemoryData for pattern.cc

TheGreatMcPain committed:
Instead of copying the 'readMemoryData' function..

simply just set it to public in 'memory.h' and use it in 'pattern.cc'

TheGreatMcPain committed:
Implement Module.getModules()

Uses similer code to findPattern, but instead of looking for a matching
string it just takes whatever's in /proc/.../maps and turns it into a
vector of modules.

TheGreatMcPain committed:
Implement findPattern in memoryjs.

TheGreatMcPain committed:
Add findPattern to index.js

TheGreatMcPain committed:
Add pattern.cc to binding.gyp

TheGreatMcPain committed:
Fix "multiple definition of `Memory'" error

TheGreatMcPain commented:
I might need some help debugging this.

The my CrewLink pull request is still "Waiting for Among Us".

I'm still pretty unfamilier with Node.js and javascript in general.

ghaith commented on lib/memoryjs.cc:
```diff
@@ -146,6 +148,60 @@ Napi::Value findModule(const Napi::CallbackInfo& args) {
+
+    char *match = strstr(moduleEntries[i].pathname, moduleName.c_str());
+
+    if (NULL == match) {
```
Shouldn't this be !=, as in moduleName is a substring of path?

TheGreatMcPain replied:
Was trying to mimic [this line](https://github.com/Rob--/memoryjs/blob/79618efdefc2aac203ae7137eefa2cf30cf888ce/lib/memoryjs.cc#L677) from the original `memoryjs` repo.

```
    if (!strcmp(moduleEntries[i].szModule, moduleName.c_str())) {
```

Then again I'm not entirely why they did it like that.

ghaith commented on lib/memoryjs.cc:
```diff
@@ -146,6 +148,60 @@ Napi::Value findModule(const Napi::CallbackInfo& args) {
+
+  pid_t hProcess = (pid_t)args[0].As<Napi::Number>().Int64Value();
+
+  std::vector<module::Module> moduleEntries = module::getModules(hProcess, &errorMessage);
```
I don't know if we are actually crashing here, but out of interest, did you try to use findModule here? since we know it works? and find the pattern on that directly?

TheGreatMcPain replied:
I haven't tried using `findModule` yet, but I'll see if it helps later today.

TheGreatMcPain replied:
I just tried `findModule()` instead of using `getModule()`.

Looks like my `findPattern()` in `pattern.cc` is not working.

TheGreatMcPain committed:
Forgot to export 'findPattern'. Oops

TheGreatMcPain committed:
Try findModule instead of getModules.

TheGreatMcPain commented:
I think my `findPattern()` function in `pattern.cc` is not completely working
yet, because CrewLink now exits immediately after it sees Among Us, and
`dmesg` is showing some segfaults for electron.

TheGreatMcPain commented:
I've setup a separate ~[GitHub repo](https://github.com/TheGreatMcPain/memoryjs-test-linux)~ that only contains the memoryjs C++ sources
with all the node.js stuff stripped out, so that I can debug the C++ code
easier as I'm more comfortable with `gdb` and friends.

The first thing I tested was my `getModules` code which shares similar code
to `findModule`. I ran into a double free issue which happened when
adding `module::Module`s to `getModules()`'s vector. I solved it by having it
return a vector of Module pointers instead of regular Modules. The downside is that
each Module pointer in the returned vector would need to be deleted.

I'm about to try `findPattern()` to see if there're any issues
there.

EDIT: I've deleted the linked repository.

TheGreatMcPain committed:
Return a vector of Module pointers in getModules.

Prevents potential 'double free()' errors.

TheGreatMcPain committed:
Attempt to use getModules in findPattern again.

TheGreatMcPain committed:
Try to parse all the lines in /proc/.../maps

ghaith commented:
I've been trying to debug the findPattern using your repo
I used this in main
```cpp

  //Get the Among us process
  //Imitate the find pattern call
  const char* errorMessage = "";
  auto pid = Process.openProcess("Among Us.exe", &errorMessage);
  std::cout << "Find Module" << std::endl;
  module::Module mod = module::findModule("GameAssembly.dll", pid , &errorMessage);

  auto sig = "48 8B 05 ? ? ? ? 48 8B 88 ? ? ? ? 48 8B 01 48 85 C0 0F 84 ? ? ? ? 66 66 66 0F 1F 84 00 ? ? ? ?";
  std::cout << "Find Pattern" << std::endl;
  uintptr_t res = Pattern.findPattern(pid,mod, sig,  0 | 2 ,  3,  4);
  sig = "48 8B 05 ? ? ? ? 48 8B 88 ? ? ? ? 74 72 48 8B 39 48 8B 0D ? ? ? ? F6 81 ? ? ? ? ?";
  std::cout << res << std::endl;
  std::cout << "Find Pattern" << std::endl;
  res = Pattern.findPattern(pid,mod, sig,  0 | 2 ,  3,  4);
  sig = "48 8B 05 ? ? ? ? 48 8B 88 ? ? ? ? 48 8B 01 48 85 C0 0F 84 ? ? ? ? BE ? ? ? ?";
  std::cout << res << std::endl;
  std::cout << "Find Pattern" << std::endl;
  res = Pattern.findPattern(pid,mod, sig,  0 | 2 ,  3,  4);
  std::cout << res << std::endl;
```

It looks like once we read the module memory in findPattern, the array does not contain any of the patterns we are looking for. Not sure how to debug further, so i'm putting this here incase someone has a clue.

I got the signatures from the CrewLink sources. These are the actual values that get called in Crewlink

TheGreatMcPain commented:
So, I did a little digging through the CrewLink code, and noticed that one
of the address offset for `meetingHud` is `0x1c573a4` in hex (29717412 in base 10).

If we use a base address of `0x3230000` and add the offset for `meetingHud`
we would get `0x4e873a4`. Based on the `/proc/.../maps` that I looked at
while Among Us was running. The start and end address for `GameAssembly.dll`
was only `0x3230000` and `0x3231000` which is nowhere `0x4e873a4`.

So, I may need to tweak `getModule` so that we can properly search all the
addresses that `GameAssembly.dll` has.

TheGreatMcPain committed:
Make 'Module.pathname' a static array.

I had a lot of trouble trying to debug a 'double free()' while working
on 'getModules' along with some random memory corruption.

TheGreatMcPain committed:
getModules: Give all modules a pathname.

This allows findPattern to search all possibly relevant memory addresses
for a given module name.

Also, have getModules return Modules instead of Module pointers.

TheGreatMcPain committed:
Have findPattern search all modules that contain the module's name.

TheGreatMcPain commented:
Okay, so I'm at the point where `findPattern` will return something that's not `-2` using the values provided by CrewLink's sourcecode.

Unfortunately CrewLink still doesn't work. My guess it that `findPattern` doesn't get the correct values dispite it finding the requested pattern.

I think I'm gonna try to get the correct addresses from the Windows version, so that we actually know what the correct offsets are.

I might just use the values from the Windows version of CrewLink to see if I can just supply bypass `findPattern` in CrewLink.

TheGreatMcPain commented:
Okay, so I'm at the point where `findPattern` will return something that's not `-2` using the values provided by CrewLink's sourcecode.

Unfortunately CrewLink still doesn't work. My guess it that `findPattern` doesn't get the correct values dispite it finding the requested pattern.

TheGreatMcPain committed:
findPattern: Substract result address with a given base address.

This should give CrewLink the correct offsets since they were wrong
before.

TheGreatMcPain commented:
I think this is ready to go unless @zbanks has some suggestions.

TheGreatMcPain committed:
getModules: skip lines containing '['.

This skips things like '[heap]' and '[stack]' which we probably don't
need right now.

Also, use 'strchr' where we're only searching for one character.

TheGreatMcPain committed:
module.cc: Minor formatting.
